mesonpy/__init__.py

```diff
@@ -126,6 +128,43 @@ def _init_colors() -> Dict[str, str]:
 assert all(re.match(_EXTENSION_SUFFIX_REGEX, x) for x in _EXTENSION_SUFFIXES)
 
 
+# Maps wheel installation paths to Meson installation path placeholders.
+_SCHEME_MAP = {
+    'scripts': ('{bindir}',),
+    'purelib': ('{py_purelib}',),
+    'platlib': ('{py_platlib}', '{moduledir_shared}'),
+    'headers': ('{includedir}',),
+    'data': ('{datadir}',),
+    # our custom location
+    'mesonpy-libs': ('{libdir}', '{libdir_shared}')
+}
+
+
+def _map_meson_destination(destination: str) -> Tuple[Optional[str], pathlib.Path]:
+    """Map Meson installation path placeholders to wheel installation directory."""
+    parts = pathlib.Path(destination).parts
+    for folder, placeholders in _SCHEME_MAP.items():
+        if parts[0] in placeholders:
+            return folder, pathlib.Path(*parts[1:])
+    return None, pathlib.Path(destination)
+
+
+def _map_to_wheel(sources: Dict[str, Dict[str, Any]]) -> DefaultDict[str, List[Tuple[pathlib.Path, str]]]:
+    """Map files to the wheel, organized by wheel installation directrory."""
+    wheel_files = collections.defaultdict(list)
+    for group in sources.values():
+        for src, target in group.items():
+            dest = target['destination']
+            directory, path = _map_meson_destination(dest)
+            if directory is not None:
+                wheel_files[directory].append((path, src))
+            else:
+                warnings.warn(f'Installation path {dest!r} for {src!r} could not be mapped to an equivalent wheel directory.')
+                if not re.match(r'^{\w+}$', path.parts[0]):
+                    raise RuntimeError('Meson installation path {dest!r} does not start with a placeholder. Meson bug!')
+    return wheel_files
```

Comment:
> Please keep these functions in `_WheelBuilder`.

Why? They don't need access to any member of `_WheelBuilder`.

> IMO it'd be cleaner to move the warnings to `_map_meson_destination`.

For doing that, either the src path needs to be passed to `_map_meson_destination` just for the purpose of maybe emitting the warning this information needs to be removed from the warning.

Reply:
> Why? They don't need access to any member of `_WheelBuilder`.

Because its whole purpose is to support code in `_WheelBuilder`, and this way it also can access internal state, for example the source paths you referred in the next comment.

> For doing that, either the src path needs to be passed to `_map_meson_destination` just for the purpose of maybe emitting the warning this information needs to be removed from the warning.

Or you can have it in `_WheelBuilder` and be able to access it directly :sweat_smile:

Reply:
Please have a look at how the code is written. Having it a method of `_WheelBuilder` does not help in any way.